dash-renderer/src/TreeContainer.js

```diff
@@ -155,6 +147,14 @@ class BaseTreeContainer extends Component {
                     })
                 );
             }
+
+            // Always update this component's props
+            _dashprivate_dispatch(
+                updateProps({
+                    props: changedProps,
+                    itempath: _dashprivate_path
+                })
+            );
```

Comment:
Callbacks are processed after a `0ms` timeout - updating the props after triggering the notifications ensures the notifications are ready on re-render.

Reply:
Just to make sure I'm understanding this correctly: `notifyObservers` is what ultimately triggers the callback processing and attaches the `rendered` event listener. This is async (with a zero delay), but so is the rerender caused by `updateProps`, and by reversing the order here we ensure that the event listener is inserted in the queue before the rerender so we're in the clear.

Do I have that right?

It still feels a bit fragile, like what we ideally would want to do is attach the event listener synchronously here somehow and just let it be used later, in case another `0ms` delay got added unexpectedly. But that might be invasive, and anyway you've got a very clear test - very nice collection of new test components!

Reply:
Yes about the ordering change. TBH I have the same discomfort about the fragility or at least the fragmentation of the implementation between the APIController `useEffect`, https://github.com/plotly/dash/blob/dev/dash-renderer/src/actions/isAppReady.js#L38 DOM check and the async wait in `observers/requestedCallbacks`. What's certain is that we want the event to be triggered when we are certain the DOM has been updated after a React render, otherwise the DOM check might be moot, and that pretty much leaves only useEffect/componentDidUpdate/componentDidMount.

I've tested adding an additional `0ms` async delay prior to calling `notifyObservers` and it does fail the test, so a change in timing would probably be caught, which is comforting..